### PR TITLE
mix appsignal.diagnose

### LINKED ISSUE #47: mix appsignal.diagnose

### PR DESCRIPTION
**Pull request by tombruijn**
As currently present in our Ruby gem as well as "appsignal diagnose".

The "appsignal.diagnose" command is useful for debugging AppSignal and
the configuration of AppSignal in an application.

Debugging manual referencing the "diagnose" command:
http://docs.appsignal.com/support/debugging.html

Closes #47 

## TODO

I created pending specs for the following:

- [x] Implement `root` user detection. Is the user that started the elixir process the `root` user? Ruby version: https://github.com/appsignal/appsignal-ruby/blob/d7f31875aaa9f4227eca46019692b4e8595b8325/spec/lib/appsignal/cli/diagnose_spec.rb#L93-L111
- [x] Implement Nif loaded detection. See #95 Ruby version: https://github.com/appsignal/appsignal-ruby/blob/d7f31875aaa9f4227eca46019692b4e8595b8325/spec/lib/appsignal/cli/diagnose_spec.rb#L49-L80
- [x] Implement "path" ownership detection. Who is owner of the paths we test and do they match with the user that has started the elixir process? Ruby version: https://github.com/appsignal/appsignal-ruby/blob/d7f31875aaa9f4227eca46019692b4e8595b8325/spec/lib/appsignal/cli/diagnose_spec.rb#L281-L323
- [ ] Test output from agent in diagnostics mode as seen here: https://github.com/appsignal/appsignal-ruby/blob/0e97b6f9180c4504e4361021fafa6ec7da3e6d42/spec/lib/appsignal/cli/diagnose_spec.rb#L55-L64 Looks like the output from a C-extension is not being captured by ExUnit's `capture_io`? It's also not outputted in test test itself. So something is hiding it? Ruby version: https://github.com/appsignal/appsignal-ruby/blob/d7f31875aaa9f4227eca46019692b4e8595b8325/spec/lib/appsignal/cli/diagnose_spec.rb#L55-L64
- [x] Heroku detection
- [ ] Docker detection (can wait until later, see also agent issue: https://github.com/appsignal/appsignal-agent/issues/212)

### CODE REVIEW

File: lib/appsignal/system.ex [hunk 1]
```diff
@@ -2,4 +2,11 @@ defmodule Appsignal.System do
   def heroku? do
     System.get_env("DYNO") != nil
   end
+
+  def root? do
+    case System.cmd("id", ["-u"]) do
```
Comment (arjan) [inline lib/appsignal/system.ex]: `System.get_env("USER") == "root"` is also an option.

Comment (tombruijn) [inline lib/appsignal/system.ex]: For this scenario, probably, assuming the `root` user is always named `root`.

But the `uid` is also used for another part of the diagnose command. In the "Paths" section we compare the owner of paths and files in the "paths" section to debug the ownership of the "appsignal paths". That's why we need to know the `uid` number. `Appsignal.System.root?/0` can remain, but we also need a more direct function to receive the `uid`, such as `Appsignal.System.uid/0`

Side-note: Not finding a good way to uncover the `uid` for the process I added the `Process user` entry to the diagnose output which does just that. Now that we've decided on a way to fetch the `uid` from the system by shelling out, the "Process user" output line can be removed.

**Comment by jeffkreeftmeijer** [comment]
To check if the NIF is loaded, we first need to implement #90, and build a way to know if it's loaded. I'm okay with merging like this, and adding another issue to make sure this task gets fixed after implementing #90.

**Comment by jeffkreeftmeijer** [comment]
@arjan do you have some insight into Tom's "Test output from agent in diagnostics"-todo?

**Comment by tombruijn** [comment]
I've marked the test for "when Nif is not loaded" as pending, so it won't break the build.

I'm also okay with merging it without that NIF loaded check. Same for docker detection. I would only be really sad if there's no way to test/capture the agent diagnostic mode output.

**Comment by tombruijn** [comment]
@arjan about the "Test output from agent in diagnostics" item. Do you have any idea how to capture this? At first I thought ExUnit's `capture_io` doesn't capture this output, but after more testing I haven't been able to see the agent get started with `Application.ensure_all_started(:appsignal)` in this test. I've tried manually starting the Nif `Appsignal.Nif.start`, making sure it's stopped beforehand `Appsignal.Nif.stop`, but I don't see it show up as a process so I'm starting to think I can't start it this way. Any ideas?

The test in question is the test marked with the `pending` tag

**Comment by tombruijn** [comment]
I think we can merge it as right now, maybe a rebase after #104 is merged. We'll mark the "test output from agent in diagnostics"  test as pending (as it is now) and return to it at a later time.

agreed @jeffkreeftmeijer ?

**Pull request merged**